docs/matrix.md

```diff
@@ -155,6 +155,84 @@ Combinations generated
 { "IMAGE": "image-3", "DOCKERFILE": "path/to/Dockerfile3}
 ```
 
+## DisplayName
+
+Matrix creates multiple `taskRuns` with the same `pipelineTask`. Each `taskRun` has its unique combination `params` based
+on the `matrix` specifications. These `params` can now be surfaced and used to configure unique name of each `matrix`
+instance such that it is easier to distinguish all the instances based on their inputs.
+
+```yaml
+pipelineSpec:
+  tasks:
+    - name: platforms-and-browsers
+      displayName: "Platforms and Browsers: $(params.platform) and $(params.browser)"
```

Comment:
Should this be `matrix.params.platform` and `matrix.params.browser`?

It looks like this wasn't included in the final version of the proposal that was merged, but I wonder if that was an oversight as it looked like it was agreed in the review feedback.

i.e. that there would be matrix-specific variables for the substitution to differentiate from other params, e.g.: https://github.com/tektoncd/community/pull/1122#discussion_r1452386634

This would be important to allow adding support for pipelineTask param substitution in future.

Maybe I missed some other discussion that changed this plan?

Reply:
thanks @AlanGreene for following up!

We had a discussion in the API WG. A keyword `matrix` in `params.platform` is not needed as the `displayName` is now part of `childReferences` irrespective of whether a task is a `matrix` or not.

With this approach, the pipelineTask param substitution in displayName is also done in this PR.

Do you think this could cause any issues while the clients are trying to rely on this change?

Reply:
> Is it possible to have a matrix param with the same name as a pipelineTask / implicitly propagated param? 

Hey @AlanGreene, matrix params has to be named same as the task params and implicitly mapped to the task params. In the example of platform and browsers, both those parameters are mapped to those params of the referenced task.

It is possible that the pipeline author can include matrix params which does not exist in the referenced task and hence will be missing from the `taskRun`. But that can create other issues before getting to the `displayName`. If not then the displayName cannot be resolved as they will not be present in the `taskRun`.

Reply:
I think I understand now what you are referring to 😥 

It is not allowed, we have a validation in place to make sure the same param either exist in the matrix or as pipelineTask param.

```
    tasks:
      - name: platforms-and-browsers
        displayName: "Platform: $(params.platform) with Browser: $(params.browser)"
        params:
          - name: platform
            value: linux1
        matrix:
          params:
            - name: platform
              value:
                - linux
```

results in:

```
 k create -f examples/v1/pipelineruns/beta/pipelinerun-with-matrix.yaml 
task.tekton.dev/platform-browsers created
Error from server (BadRequest): error when creating "examples/v1/pipelineruns/beta/pipelinerun-with-matrix.yaml": admission webhook "validation.webhook.pipeline.tekton.dev" denied the request: validation failed: expected exactly one, got both: spec.pipelineSpec.tasks[0].matrix[platform], spec.pipelineSpec.tasks[0].params[platform]
```

Reply:
OK that's good to know, however there may still be an issue with propagated params, see for example:

```yaml
apiVersion: tekton.dev/v1
kind: PipelineRun
metadata:
  generateName: pr-echo-
spec:
  params:
    - name: HELLO
      value: "Hello World!"
  pipelineSpec:
    tasks:
      - name: echo-hello
        displayName: Message is $(params.HELLO)
        matrix:
          params:
            - name: HELLO
              value:
                - hi
                - hey
        taskSpec:
          steps:
            - name: echo
              image: ubuntu
              script: |
                #!/usr/bin/env bash
                echo "$(params.HELLO)"
```

This is accepted and executed by the controller today, however it does not produce the expected result. The matrix params are correctly applied to the TaskRun params, however the params in the step definition have already been replaced by the pipeline params.

This seems like a bug, or at least an undocumented limitation, with the current matrix implementation.

How would this scenario be handled with the new displayName processing?

Reply:
Conflicts between task/matrix parameters and pipeline parameters are indeed a possibility. 
If anything, I believe we should give precedence to local (matrix) parameters in case of naming conflict, as it is more likely that the author would like to use those in the `displayName` rather than a fixed pipeline parameter.

That would be an acceptable compromise for me, which would allow us to avoid introducing the `matrix` prefix

Reply:
Matrix params taking precedence makes sense 👍 Of course it should be consistent between the displayName and actual usage in the steps.

Do we need a separate issue for the bug illustrated by the example above where param substitution in the steps is NOT using the matrix params?

Reply:
Posting here to record offline discussion
- option 1 sounds good to me
- and we'll file a bug to track the bad substitution for taskSpec

Reply:
Hey Alan, I have created an issue here - https://github.com/tektoncd/pipeline/issues/7704

Reply:
Marking this conversation as resolved - thanks @pritidesai for issue #7704